File: challenges/999/916.WordSubsets.go
package challenges

/**
We are given two arrays A and B of words.  Each word is a string of lowercase letters.

Now, say that word b is a subset of word a if every letter in b occurs in a, including multiplicity.  For example, "wrr" is a subset of "warrior", but is not a subset of "world".

Now say a word a from A is universal if for every b in B, b is a subset of a.

Return a list of all universal words in A.  You can return the words in any order.



Example 1:

Input: A = ["amazon","apple","facebook","google","leetcode"], B = ["e","o"]
Output: ["facebook","google","leetcode"]
Example 2:

Input: A = ["amazon","apple","facebook","google","leetcode"], B = ["l","e"]
Output: ["apple","google","leetcode"]
Example 3:

Input: A = ["amazon","apple","facebook","google","leetcode"], B = ["e","oo"]
Output: ["facebook","google"]
Example 4:

Input: A = ["amazon","apple","facebook","google","leetcode"], B = ["lo","eo"]
Output: ["google","leetcode"]
Example 5:

Input: A = ["amazon","apple","facebook","google","leetcode"], B = ["ec","oc","ceo"]
Output: ["facebook","leetcode"]


Note:

1 <= A.length, B.length <= 10000
1 <= A[i].length, B[i].length <= 10
A[i] and B[i] consist only of lowercase letters.
All words in A[i] are unique: there isn't i != j with A[i] == A[j].
*/

func wordSubsets(a []string, b []string) []string {
	cnt := make([]int, 26)

	for _, w := range b {
		chars := make([]int, 26)
		for _, c := range w {
			chars[int(c-'a')]++
		}

		for i := range cnt {
			if chars[i] > cnt[i] {
				cnt[i] = chars[i]
			}
		}
	}

	ans := make([]string, 0, len(a))

	for _, w := range a {
		chars := make([]int, 26)
		for _, c := range w {
			chars[int(c-'a')]++
		}

		found := true
		for i := range chars {
			if chars[i] < cnt[i] {
				found = false
				break
			}
		}

		if found {
			ans = append(ans, w)
		}
	}

	// fmt.Println(cnt)

	return ans
}

File: challenges/999/916_WordSubsets.py
'''
You are given two string arrays words1 and words2.

A string b is a subset of string a if every letter in b occurs in a including multiplicity.

For example, "wrr" is a subset of "warrior" but is not a subset of "world".
A string a from words1 is universal if for every string b in words2, b is a subset of a.

Return an array of all the universal strings in words1. You may return the answer in any order.

Example 1:

Input: words1 = ["amazon","apple","facebook","google","leetcode"], words2 = ["e","o"]
Output: ["facebook","google","leetcode"]
Example 2:

Input: words1 = ["amazon","apple","facebook","google","leetcode"], words2 = ["l","e"]
Output: ["apple","google","leetcode"]

Constraints:

1 <= words1.length, words2.length <= 10^4
1 <= words1[i].length, words2[i].length <= 10
words1[i] and words2[i] consist only of lowercase English letters.
All the strings of words1 are unique.
'''

from typing import List
from collections import defaultdict, Counter


class Solution:
  def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
    full_set = defaultdict(int)
    
    for w in words2:
      c = Counter(w)
      for ch, cnt in c.items():
        full_set[ch] = max(full_set[ch], cnt)
      
    ans = []
    for w in words1:
      c = Counter(w)
      found = True
      
      for ch in full_set:
        if ch not in c or c[ch] < full_set[ch]:
          found = False
          break
          
      if found:
        ans.append(w)
      
    return ans